### PR TITLE
multi-server: jobs which span boundaries stay queued when it is eligible to run

### PR DESCRIPTION
**Pull request by nithinj**
#### Describe Bug or Feature
multi-server: jobs which span boundaries stay queued when it is eligible to run as initially servers doesn't have the node cache. It will get in the next scheduling cycle.

#### Describe Your Change
Server will fetch the node cache when it comes up.
Server will inform the scheduler once it receives the cache to trigger another cycle.

#### Attach Test and Valgrind Logs/Output

[test.log](https://github.com/openpbs/openpbs/files/6541716/test.log)

Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7699|4178|1|5|0|15|4157|

Description: Rerun Only Failed Tests From #7699
Platforms: UBUNTU1804,SLES12
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7712|6|0|0|0|0|6|

### CODE REVIEW

File: src/server/multi_svr.c [hunk 1]
```diff
@@ -1029,6 +1032,10 @@ update_node_cache(int stream, struct batch_status *bstat)
 
 	init_node_from_bstat(bstat, psvr);
 
+	/* Trigger default scheduler to schedule the job which has failed due
+	to missing vnode cache information */
+	set_scheduler_flag(SCH_SCHEDULE_NEW, dflt_scheduler);
```
Comment (agrawalravi90) [inline src/server/multi_svr.c:1037]: why only default? I think server should trigger sched cycles for all schedulers that it knows about, otherwise there can be jobs on other scheds which the default sched will ignore

File: src/server/multi_svr.c [hunk 2]
```diff
@@ -613,8 +613,11 @@ connect_to_peersvr(void *psvr)
 	if (send_connect(psvr) < 0)
 		return -1;
 
-	if (resc_upd_reqd && svr_info->ps_pending_replies)
-		mcast_resc_update_all(psvr);
+	if (resc_upd_reqd) {
+		if (svr_info->ps_pending_replies)
+			mcast_resc_update_all(psvr);
+		send_nodestat_req();
```
Comment (agrawalravi90) [inline src/server/multi_svr.c:619]: Shouldn't we remove the other calls to send_nodestat_req()? are they still needed?

File: src/server/svr_comm.c [hunk 3]
```diff
@@ -271,15 +271,15 @@ send_nodestat_req(void)
 
 	update_msvr_stat(1, CACHE_MISS);
 
+	if (mtfd == -1)
+		return;
```
Comment (agrawalravi90) [inline src/server/svr_comm.c:275]: shouldn't this be done before the call to update_msvr_stat() ?

**Review by agrawalravi90 — CHANGES_REQUESTED** [review]

File: src/server/multi_svr.c (hunk 1 above)
Comment (nithinj) [reply src/server/multi_svr.c:1037]: We should think about whether we need to support multi-sched in its current form with multi-svr. We can revisit this if we decide to do so.

File: src/server/multi_svr.c (hunk 2 above)
Comment (nithinj) [reply src/server/multi_svr.c:619]: yes, send_nodestat_req() gets invoked when node cache becomes invalid

File: src/server/svr_comm.c (hunk 3 above)
Comment (nithinj) [reply src/server/svr_comm.c:275]: update_msvr_stat() is updating a potential cache miss. that has to be done anyway except when it is initially invoked. I'll add the changes for that.

File: src/server/svr_comm.c [hunk 4]
```diff
@@ -257,29 +259,30 @@ replyhello_psvr(struct work_task *ptask)
...
 	struct attrl *head;
 	struct attrl *tail;
 	struct attrl *nxt;
 	int mtfd = open_ps_mtfd();
-	int rc;
 	static int time_last_sent = 0;
+	int rc;
 
-	update_msvr_stat(1, CACHE_MISS);
+	update_msvr_stat(1, type);
```
Comment (ashwathraop) [inline src/server/svr_comm.c:275]: All the calls of send_nodestat_req() take CACHE_MISS, so, why did you change this to access CACHE_MISS through `type` variable?

Comment (nithinj) [reply src/server/svr_comm.c:275]: type can also be -1. Which is handled within update_msvr_stat instead of the caller.

**Review by agrawalravi90 — APPROVED** [review]
LGTM

**Review by ashwathraop — APPROVED** [review]
